pkg/config/expand.go

```diff
@@ -162,57 +162,79 @@ func getModuleInputMap(inputs []modulereader.VarInfo) map[string]string {
+	_, found := mod.Settings[settingName]
+	if !found {
+		mod.Settings[settingName] = []interface{}{}
+		mod.createWrapSettingsWith()
+		mod.WrapSettingsWith[settingName] = []string{"flatten(", ")"}
+	}
+	currentValue, ok := mod.Settings[settingName].([]interface{})
+	if ok {
+		mod.Settings[settingName] = append(currentValue, value)
+		return nil
+	}
+	return fmt.Errorf("%s: module %s, setting %s",
+		errorMessages["appendToNonList"], mod.ID, settingName)
+}
+
+// This function matches input variables in a "using" module to output values
+// from a "used" module. It may be used iteratively to successively apply used
+// modules in order of precedence. New input variables are added to the using
+// module as Toolkit variable references (in same format as a blueprint). If
+// the input variable already has a setting, it is ignored, unless the value is
+// a list, in which case output values are appended and flattened using HCL.
+// mod: "using" module as defined above
+// useMod: "used" module as defined above
+// useModGroupID: deployment group ID to which useMod belongs
+// modInputs: input variables as defined by the using module code
+// useOutputs: output values as defined by the used module code
+// settingsToIgnore: a list of module settings not to modify for any reason;
+//
+//	typical usage will be to leave explicit blueprint settings unmodified
 func useModule(
 	mod *Module,
 	useMod Module,
 	useModGroupID string,
-	modInputs map[string]string,
+	modInputs []modulereader.VarInfo,
 	useOutputs []modulereader.VarInfo,
-	settingsInBlueprint []string,
-) (usedVars []string) {
-	usedVars = []string{}
+	settingsToIgnore []string,
+) ([]string, error) {
+	usedVars := []string{}
+	modInputsMap := getModuleInputMap(modInputs)
 	for _, useOutput := range useOutputs {
 		settingName := useOutput.Name
 
-		// Skip settings explicitly defined by users
-		if slices.Contains(settingsInBlueprint, settingName) {
+		// Explicitly ignore these settings (typically those in blueprint)
+		if slices.Contains(settingsToIgnore, settingName) {
 			continue
 		}
 
-		// Skip settings that do not match module inputs
-		inputType, ok := modInputs[settingName]
+		// Skip settings that do not have matching module inputs
+		inputType, ok := modInputsMap[settingName]
 		if !ok {
 			continue
 		}
 
-		// if the input exists in the module doing the using, then we modify it
-		// according to these rules:
-		// if the setting has a value, then it was set by "use" because we have
-		// skipped anything explicitly in the blueprint. If the setting is a
-		// list of any type, then we want to concatenate these values in order
-		// they were used; otherwise, the setting should remain set to the
-		// existing used value
 		_, setByUse := mod.Settings[settingName]
 		modVarName := getModuleVarName(useModGroupID, useMod.ID, settingName)
 		isInputList := strings.HasPrefix(inputType, "list")
+
+		if !setByUse && !isInputList {
```

Comment:
nit. The readability can be improved if we swap those two blocks and make it into if-else-if chain:
```
if isInputList {
  ...
} else if !setByUse {
  ...
}
```

Also `setByUse` is somewhat confusing nmae, removed gave an explanation, but no it rather unclear. We could remove any  assumptions of external interactions (set by another `use`) and simply name `alreadySet`.

Reply:
That happens to be the if/else/if block that I am replacing. Although I take the point, combined with above that we can do a few edits and keep the cyclomatic complexity in place. In particular clarifying that the two if blocks in sequence can never be reached in the same run.